### ISSUE TITLE
[CT-2173] [Feature] Eliminate cost for query run to detect on_schema_change

### ISSUE THREAD

**Issue by sai3563** [dbt-labs/dbt-bigquery#557]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-bigquery functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Ok so as you might know, for incremental models, dbt has an `on_schema_change` config which allows syncing columns of destination with the source and few other options.

Whenever the value of `on_schema_change` is anything other than `ignore`, DBT creates a temporary table by running the entire base query.

From the source code,

```
{% if on_schema_change != 'ignore' or language == 'python' %}
      {#-- Check first, since otherwise we may not build a temp table --#}
      {#-- Python always needs to create a temp table --#}
      {%- call statement('create_tmp_relation', language=language) -%}
        {{ declare_dbt_max_partition(this, partition_by, compiled_code, language) +
           bq_create_table_as(partition_by.time_ingestion_partitioning, True, tmp_relation, compiled_code, language)
        }}
      {%- endcall -%}
      {% set tmp_relation_exists = true %}
      {#-- Process schema changes. Returns dict of changes if successful. Use source columns for upserting/merging --#}
      {% set dest_columns = process_schema_changes(on_schema_change, tmp_relation, existing_relation) %}
    {% endif %}
```

However, this has an unintended side affect. During an incremental run, the same query is run twice. The first time, temporary table is created to detect schema change. Then again, same query is run within `merge into` to upsert data into the target.

There is some inconsistency when it comes to temp table usage in static and dynamic insert overwrite methods. Consider below example which describes static insert_overwrite method:

```
{{
    config(
        materialized = 'incremental',
        unique_key = 'id',
        on_schema_change = 'sync_all_columns',
        partition_by = {
          'field': 'created_date',
          'data_type': 'datetime',
          'granularity': 'day'
        },
        require_partition_filter = false,
        incremental_strategy = 'insert_overwrite',
        partitions = ["'2023-02-01'", "'2023-02-02'"]
    )
}}

  SELECT
    *
  FROM
    {{ source(source_schema,source_table) }}
{% if is_incremental() %}
  -- this filter will only be applied on an incremental run
    WHERE date(created_date) in ('2023-02-01', '2023-02-02')
      and updated_date >= '2023-02-20 19:00:01.873000'
{% endif %}
```

It generates query as follows:

```
create or replace table `project`.`target_dataset`.`table_name__dbt_tmp`
    partition by datetime_trunc(created_date, day)
    
    OPTIONS(
      expiration_timestamp=TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 12 hour)
    )
    as (

-- Use the `ref` function to select from other models

  SELECT
    *
  FROM
    `project`.`source_dataset`.`table_name`
  -- this filter will only be applied on an incremental run
  WHERE date(created_date) in ('2023-02-01', '2023-02-02')
      and updated_date >= '2023-02-20 19:00:01.873000'

    );


merge into `project`.`target_dataset`.`table_name` as DBT_INTERNAL_DEST
        using (

-- Use the `ref` function to select from other models

  SELECT
    *
  FROM
    `project`.`source_dataset`.`table_name`
  -- this filter will only be applied on an incremental run
  WHERE date(created_date) in ('2023-02-01', '2023-02-02')
      and updated_date >= '2023-02-20 19:00:01.873000'
) as DBT_INTERNAL_SOURCE
        on FALSE

    when not matched by source
         and datetime(datetime_trunc(DBT_INTERNAL_DEST.created_date, day)) in (
              '2023-02-01', '2023-02-02'
          )
        then delete

    when not matched then insert
        (`id`, `created_date`, `updated_date`, `status`)
    values
        (`id`, `created_date`, `updated_date`, `status`)
```

As you can see, even though temp table is created at first, it is not used later on within `merge into`.

This is what the query log looks like for dynamic insert_overwrite:

```
    create or replace table `project`.`target_dataset`.`table_name__dbt_tmp`
    partition by datetime_trunc(created_date, day)
    
    OPTIONS(
      expiration_timestamp=TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 12 hour)
    )
    as (


-- Use the `ref` function to select from other models

  SELECT
    *
  FROM
    `project`.`source_dataset`.`table_name`
  -- this filter will only be applied on an incremental run
  WHERE date(created_date) in ('2023-02-01', '2023-02-02')
      and updated_date >= '2023-02-20 19:00:01.873000'

    );
   
      -- generated script to merge partitions into `project`.`target_dataset`.`table_name`
      declare dbt_partitions_for_replacement array<datetime>;

      
      
        -- 1. temp table already exists, we used it to check for schema changes
      

      -- 2. define partitions to update
      set (dbt_partitions_for_replacement) = (
          select as struct
              array_agg(distinct datetime(datetime_trunc(created_date, day)))
          from `project`.`target_dataset`.`table_name__dbt_tmp`
      );

      -- 3. run the merge statement
      

    merge into `project`.`target_dataset`.`table_name` as DBT_INTERNAL_DEST
        using (
        select
        * from `project`.`target_dataset`.`table_name__dbt_tmp`
      ) as DBT_INTERNAL_SOURCE
        on FALSE

    when not matched by source
         and datetime(datetime_trunc(DBT_INTERNAL_DEST.created_date, day)) in unnest(dbt_partitions_for_replacement) 
        then delete

    when not matched then insert
        (`id`, `created_date`, `updated_date`, `status`)
    values
        (`id`, `created_date`, `updated_date`, `status`)

;

      -- 4. clean up the temp table
      drop table if exists `project`.`target_dataset`.`table_name__dbt_tmp`
```

Here, first temp table is created for schema comparison, it is then used later on within `merge into`.

In both methods, we are charged twice per job run. BigQuery doesn't use caching in this case because caching is only used when query run is exactly the same and the base table hasn't updated. That is not the case here as the query is first run under SELECT but later on, it is run within MERGE INTO.

**This increases the overall cost incurred to BigQuery due to DBT, especially at scale.**

### Describe alternatives you've considered

By appending `LIMIT 0` to the query which is run for schema comparison, we can essentially eliminate the BigQuery cost incurred in that step as BigQuery doesn't scan any data when LIMIT 0 is appended to a query. This way, we get an empty table with the latest columns, which can then be used for schema comparison.

For static insert_overwrite method, since the temp table is not used later on, we can simply append `LIMIT 0` to the schema comparison query and the model will run just fine.

But for dynamic insert_overwrite method, since the temp table is used later on within `merge into`, we would need to modify the code to use the base query instead of the temp table within `merge into`. That way, even dynamic insert_overwrite models would work.

**I've modified the source code in my local setup and confirmed that this works.**

### Who will this benefit?

Everyone using dbt-bigquery and incremental models, as it will half the cost incurred while querying source table in an incremental model.

### Are you interested in contributing this feature?

Yes, as I've said, I have already made changes in my local setup for this to work

### Anything else?

Will raise PR as soon as I get confirmation from someone to proceed.

**Comment by dbeatty10** [comment]
Good eyes @sai3563 🤩 !

Do you think this is the same as https://github.com/dbt-labs/dbt-bigquery/issues/427? Or does it look like a different issue to you?

**Comment by sai3563** [comment]
@dbeatty10 Thanks for the quick response!!!

I would say this is different.  #427 raises a valid point that temp table created during static insert_overwrite method isn't used later on but it is on in dynamic insert_overwrite, as I've described above. But it doesn't eliminate cost incurre like my method does.

Lets say your base query scans 5 GB of data to generate a temp table of 1 GB. Then, during the schema comparison query, you would be charged 5 GB and during `merge into`, you would be charged 1 GB **for the source table**. Total would be 6 GB.

My method proposes adding `LIMIT 0` before `bq_static_insert_overwrite_sql` is called i.e. during schema comparison. That would completely eliminate cost incurred for schema check. So for the above example, you would only be charged 5 GB instead of 6. Of course, based on the ratio of output table size/input table size, cost savings would differ.

After my implementation, #427 would essentially become redundant. Hope this clears things up. Please confirm whether to proceed.

**Comment by dbeatty10** [comment]
> That would completely eliminate cost incurred for schema check

This is compelling @sai3563  💰 

Yes, please proceed with a PR, and thanks for checking 👍

--- 

I'd still love to see https://github.com/dbt-labs/dbt-bigquery/issues/427 solved by someone who is familiar with all logic related to `bq_static_insert_overwrite_sql `.

No pressure, but it would be amazing if you'd be willing to tackle it in addition to the other two PRs that you're already offering to work on. It might be a pretty small PR (see below).

@jtcohen6 described what would need to happen [here](https://github.com/dbt-labs/dbt-bigquery/issues/427#issuecomment-1404805424):

> So I think all we'd need to fix this is, have the "static" version start accepting the `tmp_relation_exists` argument, and update this conditional logic to use it, instead of just repeating `{{ sql }}`:
> https://github.com/dbt-labs/dbt-bigquery/blob/38fb796bf213217220ed4bcc6ce237b0b25d507c/dbt/include/bigquery/macros/materializations/incremental_strategy/insert_overwrite.sql#L58-L66

**Comment by sai3563** [comment]
@dbeatty10 I have no problem solving #427, but my solution and #427 are mutually exclusive. I suggest using the full query with `LIMIT 0` for cost free schema comparison, then again using the full query to insert data into destination within `merge into`. #427 suggests using temp table within `merge into`, which wouldn't after my modification of `LIMIT 0` as the table would have no data.

Ultimately both PRs cannot be merged into one stable branch. As I've said above, my method would end up saving more cost compared to #427.

But since it a small change, I will raise PR for #427 too. 😊

**Comment by github-christophe-oudar** [comment]
I opened https://github.com/dbt-labs/dbt-bigquery/pull/630 to fix #427
@sai3563 feel free to have a look
